pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+from ast import AST, FunctionDef, unparse
+from configparser import ConfigParser
+from contextlib import contextmanager
+from dataclasses import dataclass
+from enum import IntEnum
+from inspect import FrameInfo, stack
+from typing import Callable, cast
+import os
+import re
+
+from executing import Source  # type: ignore
+
+
+@dataclass
+class StackFrame:
+    """
+    StackFrame is an _internal_ PyTeal class and is
+    the first ancestor in the following linear class hierarchy:
+                        StackFrame
+                            |
+                        PyTealFrame
+                            |
+    compiler.sourcemap.TealMapItem
+
+    Somewhat confusingly, NatalStackFrame is _NOT_ in this hierarchy
+    so not "is-a" StackFrame, but it pretends to be and "has-a" list of StackFrame's.
+
+    Its most important members are:
+        * frame_info of type inspect.FrameInfo
+        * node of type ast.AST
+    The first is a python representation of a stack frame, and the
+    second represents a python AST node. The imported package `executing`
+    features the method `Source.executing` which when run shortly after
+    the frame_info's creation, _usually_ succeeds in recovering the
+    associated AST node.
+
+    In the current usage, there is a `creator` member which is a `NatalStackFrame`
+    object -usually belonging to a PyTeal Expr- and which is assumed to have called
+    the private constructor `_init_or_drop()`.
+
+    It is not recommended that this class be accessed directly.
+
+    Of special note, is the class variable `_internal_paths`.
+    This is a whitelist of file patterns which signal to the logic
+    of method `_frame_info_is_pyteal()` that a particular
+    frame was _NOT_ created by the user.
+
+    QUIRKS:
+    * Unfortunately, this means that if a user created file satisfies the above pattern,
+    the performance of the source mapper will degrade.
+    * User generated code that includes the pattern `NatalStackFrame` may produce degraded results
+    """
+
+    frame_info: FrameInfo
+    node: AST | None
+    creator: "NatalStackFrame"
+
+    # for debugging purposes:
+    full_stack: list[FrameInfo] | None = None
```

Comment:
yeah, `full_stack` is only available when `debug` is True... this can be cleaned up

Reply:
seems we are all clear about what this value is for, just one question: "clean up" means we are going to remove this debugging purpose var in #676 or later PR?

Reply:
It can be cleaned up potentially, because instead of maintaining the full stack from the very beginning, we could just keep the origin and recreate as needed. This kind of approach wasn't always successful, however, with the lighter approach that I want to try, it may be feasible. Here's an item that I dded to #677 
<img width="649" alt="image" src="https://user-images.githubusercontent.com/291133/221611011-f45b5ffd-24e9-415c-9c54-522203441bf3.png">

Reply:
sounds good, that makes sense.